src/gallium/drivers/lima/lima_resource.c

```diff
@@ -368,13 +380,61 @@ lima_transfer_map(struct pipe_context *pctx,
    ptrans->usage = usage;
    ptrans->box = *box;
    ptrans->stride = res->stride;
+   trans->res = res;
 
    *pptrans = ptrans;
 
-   return bo->map + box->z * ptrans->layer_stride +
-      box->y / util_format_get_blockheight(pres->format) * ptrans->stride +
-      box->x / util_format_get_blockwidth(pres->format) *
-      util_format_get_blocksize(pres->format);
+   if (res->tiled) {
+      uint32_t box_x1, box_y1, box_x2, box_y2;
+      uint32_t box_start_x, box_start_y;
+      bool needs_load = false;
+
+      /* No direct mappings of tiled, since we need to manually
+       * tile/untile.
+       */
+      if (usage & PIPE_TRANSFER_MAP_DIRECTLY)
+         return NULL;
+
+      if (usage & PIPE_TRANSFER_READ)
+         needs_load = true;
+
+      box_start_x = ptrans->box.x & 15;
+      box_start_y = ptrans->box.y & 15;
+
+      /* TODO: load only borders, not whole box */
+      if (box_start_x || box_start_y)
+         needs_load = true;
+
+      if (((ptrans->box.x + ptrans->box.width) & 15) ||
+          ((ptrans->box.y + ptrans->box.height) & 15))
+         needs_load = true;
+
+      /* Align box to tile boundaries */
+      box_x1 = align(ptrans->box.x, 16);
+      box_y1 = align(ptrans->box.y, 16);
+      box_x2 = align(box_x1 + ptrans->box.width, 16);
+      box_y2 = align(box_y1 + ptrans->box.height, 16);
+
+      ptrans->box.x = box_x1;
+      ptrans->box.y = box_y1;
+      ptrans->box.width = box_x2 - box_x1;
+      ptrans->box.height = box_y2 - box_y1;
+      trans->map = malloc(ptrans->stride * ptrans->box.height * ptrans->box.depth);
```

Comment:
Can trans->map only contain the box size data? Currently the stride is the whole texture stride. Besides I'd like to use "staging" instead of "map" for trans. Maybe u_transfer_helper_transfer_map is a reference.

Reply:
Not really, but we can limit height. I'll change map to staging, it sounds reasonable.

Reply:
Why not? Seems u_transfer_helper_transfer_map does like this.

Reply:
OK, I'll look into u_transfer_helper_transfer_map implementation.